<a href="https://colab.research.google.com/github/sammanfatima/Her-ML-Journey/blob/main/Feature_Encoding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Feature Encoding
* Feature encoding is the process of converting categorical (text) data into numerical data so that a machine learning model can understand it.

# Popular method of Encoding
* Label encoding
* Ordinal encoding
* One hot encoding
* Binary encoding

In [1]:
# import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
# load data
df = sns.load_dataset('tips')
df.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


# Value count of time

In [3]:
df['time'].value_counts()

,count
time,
Dinner,176
Lunch,68


* Dinner → 176 = Dinner appears 176 times in the time column.
* Lunch → 68 = Lunch appears 68 times in the time column.

# Value count of day

In [4]:
df['day'].value_counts()

,count
day,
Sat,87
Sun,76
Thur,62
Fri,19


# Value count of smoker

In [5]:
df['smoker'].value_counts()

,count
smoker,
No,151
Yes,93


# Value count of sex

In [6]:
df['smoker'].value_counts()

,count
smoker,
No,151
Yes,93


# Encode the time in labelencoder with sklearn

In [8]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
df['encoded_time']= le.fit_transform(df['time'])
df.head()


,total_bill,tip,sex,smoker,day,time,size,encoded_time
0,16.99,1.01,Female,No,Sun,Dinner,2,0
1,10.34,1.66,Male,No,Sun,Dinner,3,0
2,21.01,3.50,Male,No,Sun,Dinner,3,0
3,23.68,3.31,Male,No,Sun,Dinner,2,0
4,24.59,3.61,Female,No,Sun,Dinner,4,0


In [9]:
df['encoded_time'].value_counts() # dinner= 0, lunch=1

,count
encoded_time,
0,176
1,68


# Ordinal encoding the day column using specific order

In [14]:
from sklearn.preprocessing import OrdinalEncoder

In [17]:
oe = OrdinalEncoder(categories=[['Thur', 'Fri', 'Sat', 'Sun']])

df['encoded_day'] = oe.fit_transform(df[['day']])

df.head()

,total_bill,tip,sex,smoker,day,time,size,encoded_time,encoded_day
0,16.99,1.01,Female,No,Sun,Dinner,2,0,3.0
1,10.34,1.66,Male,No,Sun,Dinner,3,0,3.0
2,21.01,3.50,Male,No,Sun,Dinner,3,0,3.0
3,23.68,3.31,Male,No,Sun,Dinner,2,0,3.0
4,24.59,3.61,Female,No,Sun,Dinner,4,0,3.0


In [19]:
df['encoded_day'].value_counts() # 1.0= thurs, 2.0= fri, 3.0= Sat, 4.0= Sun

,count
encoded_day,
2.0,87
3.0,76
0.0,62
1.0,19


In [20]:
oe = OrdinalEncoder()

df['encoded_day'] = oe.fit_transform(df[['day']])

df.head()

,total_bill,tip,sex,smoker,day,time,size,encoded_time,encoded_day
0,16.99,1.01,Female,No,Sun,Dinner,2,0,2.0
1,10.34,1.66,Male,No,Sun,Dinner,3,0,2.0
2,21.01,3.50,Male,No,Sun,Dinner,3,0,2.0
3,23.68,3.31,Male,No,Sun,Dinner,2,0,2.0
4,24.59,3.61,Female,No,Sun,Dinner,4,0,2.0


In [21]:
df['encoded_day'].value_counts() # encode randomly

,count
encoded_day,
1.0,87
2.0,76
3.0,62
0.0,19


# One hot encoding

In [22]:
from sklearn.preprocessing import OneHotEncoder

In [24]:
ohe = OneHotEncoder()
ohe.fit_transform(df[['smoker']]).toarray()

array([[1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [0., 1.],
       [1., 0.],
       [0., 1.

In [25]:
ohe = OneHotEncoder()
ohe.fit_transform(df[['sex']]).toarray()

array([[1., 0.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [1., 0.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [1., 0.],
       [0., 1.],
       [0., 1.],
       [1., 0.],
       [0., 1.],
       [1., 0.],
       [0., 1.],
       [1., 0.],
       [0., 1.],
       [0., 1.],
       [1., 0.],
       [1., 0.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [1., 0.],
       [0., 1.],
       [0., 1.],
       [1., 0.],
       [1., 0.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [1., 0.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [1., 0.],
       [1., 0.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [0., 1.],
       [1., 0.],
       [0., 1.

# Binary Encoder

In [30]:
pip install category_encoders

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 107.9/107.9 kB 1.2 MB/s eta 0:00:00


In [38]:
df = sns.load_dataset('tips')
df.head()

,total_bill,tip,sex,smoker,day,time,size
0,16.99,1.01,Female,No,Sun,Dinner,2
1,10.34,1.66,Male,No,Sun,Dinner,3
2,21.01,3.50,Male,No,Sun,Dinner,3
3,23.68,3.31,Male,No,Sun,Dinner,2
4,24.59,3.61,Female,No,Sun,Dinner,4


In [47]:
from category_encoders import BinaryEncoder
binary_encoder = BinaryEncoder()
df_binary = binary_encoder.fit_transform(df['day'])
df_binary

,day_0,day_1,day_2
0,0,0,1
1,0,0,1
2,0,0,1
3,0,0,1
4,0,0,1
...,...,...,...
239,0,1,0
240,0,1,0
241,0,1,0
242,0,1,0
